app/scripts/metamask-controller.js

```diff
@@ -3494,34 +3492,41 @@ export default class MetamaskController extends EventEmitter {
         this._convertMnemonicToWordlistIndices(seedPhraseAsBuffer),
       );
 
+      // Scan accounts until we find an empty one
+      const { chainId } = this.networkController.state.providerConfig;
       const ethQuery = new EthQuery(this.provider);
-      accounts = await this.keyringController.getAccounts();
-      lastBalance = await this.getBalance(
-        accounts[accounts.length - 1],
-        ethQuery,
-      );
+      const accounts = await this.keyringController.getAccounts();
+      let address = accounts[accounts.length - 1];
 
-      // seek out the first zero balance
-      while (lastBalance !== '0x0') {
-        const { addedAccountAddress } =
-          await this.keyringController.addNewAccount(accounts.length);
-        accounts = await this.keyringController.getAccounts();
-        lastBalance = await this.getBalance(addedAccountAddress, ethQuery);
-      }
+      for (let count = accounts.length; ; count++) {
+        const balance = await this.getBalance(address, ethQuery);
+
+        if (balance === '0x0') {
+          // This account has no balance, so check for tokens
+          await this.detectTokensController.detectNewTokens({
+            selectedAddress: address,
+          });
 
-      const internalAccounts = this.accountsController.listAccounts();
+          const tokens =
+            this.tokensController.state.allTokens?.[chainId]?.[address];
+          const detectedTokens =
+            this.tokensController.state.allDetectedTokens?.[chainId]?.[address];
 
-      // remove extra zero balance account potentially created from seeking ahead
-      if (accounts.length > 1 && lastBalance === '0x0') {
-        const internalAccount = internalAccounts.find(
-          (account) =>
-            account.address.toLowerCase() ===
-            accounts[accounts.length - 1].toLowerCase(),
-        );
-        if (internalAccount) {
-          await this.removeAccount(internalAccount.address);
+          if (
+            (tokens?.length ?? 0) === 0 &&
+            (detectedTokens?.length ?? 0) === 0
```

Comment:
Is there a different way of writing this?  I've not seen this `?? 0 == 0` pattern before.

Reply:
I think it's the cleanest way to handle the null/undefined case.  You cannot do `!(foo?.length >= 0)` because of the fun behavior:

```
null >= 0
true
```

And because the linter forces `===` instead of `==`, I'd have to do 4 additional lines to handle null/undefined explicitly for each.